spinoso-string/src/encoded_string.rs

```diff
@@ -0,0 +1,481 @@
+use alloc::vec::Vec;
+use core::ops::Range;
+use core::slice::SliceIndex;
+
+use bstr::BStr;
+
+use crate::ascii_string::AsciiString;
+use crate::binary_string::BinaryString;
+use crate::codepoints::InvalidCodepointError;
+use crate::encoding::Encoding;
+use crate::iter::{Bytes, IntoIter, Iter, IterMut};
+use crate::ord::OrdError;
+use crate::utf8_string::Utf8String;
+
+pub enum EncodedString {
+    Ascii(AsciiString),
+    Binary(BinaryString),
+    Utf8(Utf8String),
+}
```

Comment:
> ruby `force_encoding` is gong to be .... fun

`String#force_encoding` can be implemented like this:

```rust
impl EncodedString {
    pub fn force_encoding(&mut self, encoding: Encoding) {
        if encoding == self.encoding() {
            return;
        }
        let buf = mem::replace(self.as_mut_vec(), vec![]);
        match encoding {
            Encoding::Ascii => *self = Self::Ascii(AsciiString::new(buf)),
            Encoding::Ascii => *self = Self::Binary(BinaryString::new(buf)),
            Encoding::Ascii => *self = Self::Utf8(Utf8String::new(buf)),
        }
    }
}
```

Reply:
Ah magic!

I'll leave this for a separate PR as to not add more to the scope of this massive PR as it :grimacing: